Fix async exports udefined

laskos opened the pull request:
https://phabricator.babeljs.io/T3085

sonewman commented:
:+1:

sonewman commented:
I think the problem seems to be the ordering that the code is output rather than the semantics of ensuring that hoisting does not occur.

Because the async function gets defined first then the exports get overwritten.
```javascript
var myFunction = exports.myFunction = (function () {
   //....
  return function myFunction(_x) {
    return ref.apply(this, arguments);
  };
})();

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.myFunction = undefined;
```

sonewman commented:
- Sorry, I only just realised your link already displayed the problem.

My assumption is that the line removed in this PR is there in order to align with some kind of specification that values defined as an export are not hoisted.

But I don't know if that would apply to named async functions...

The failing tests actually assume some expected behaviour, so just removing this assignment might not be right :confused:

thejameskyle commented on packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js:
```diff
@@ -247,7 +247,6 @@ export default function () {
                     if (id.isIdentifier()) {
                       addTo(exports, id.node.name, id.node);
                       init.replaceWith(buildExportsAssignment(id.node, init.node).expression);
-                      nonHoistedExportNames[id.node.name] = true;
```
I assume based on the tests results, that instead of removing this, you need to tighten up the conditions that is it run.

laskos commented:
Ok I updated the PR. The problem was, generator transformed from async function was marked as hoisted declaration, that's why it was always on top. Remove this line fixed problem. It's fine for me - functions declaration should stay on origin position.

I added normal function to test.

hzoo commented:
Ok yeah seems like that should be removed after changing the type to `FunctionExpression` in https://github.com/babel/babel/pull/2908

sonewman commented:
is there any particular reason that the exports value is hoisted and set to undefined? I couldn't seem to find the history of why that change was introduced.
